### DISCUSSION TITLE
Lexer trouble: how to really match "any-and-all" chars?

### DISCUSSION THREAD

**Discussion by metaleap** [orbitalquark/textadept#153] [Q&A] [answered]
This lexer can be assigned to a buffer just fine and is then shown as active in its statusbar, but it just shows no visual effect whatsoever... The various `myany*` are just `lexer.any`-usage permutations in the hopes that one might take effect. But none of them does:

```lua
local lexer = require('lexer')
local token = lexer.token

local lex = lexer.new('dbgbuf')
local style = lexer.styles.default .. { bold = true, italics = true, back = 0xB8B8B8 }

lex:add_rule('myany1', token(lexer.DEFAULT, lexer.any^1))
lex:add_style(lexer.DEFAULT, style)

lex:add_rule('myany2', token(lexer.DEFAULT, lexer.any))
lex:add_style(lexer.DEFAULT, style)

lex:add_rule('myany3', token(lexer.DEFAULT, lexer.any^1))
lex:add_style('myany3', style)

lex:add_rule('myany4', token(lexer.DEFAULT, lexer.any))
lex:add_style('myany4', style)

lex:add_rule('myany5', token('myany5', lexer.any^1))
lex:add_style('myany5', style)

lex:add_rule('myany6', token('myany6', lexer.any))
lex:add_style('myany6', style)

lex:add_rule('custom_whitespace', token('custom_whitespace', lexer.space^1))
lex:add_style('custom_whitespace', style)

return lex
```

**Comment by orbitalquark** [comment, marked as answer]
From https://orbitalquark.github.io/textadept/api.html#rules:

"Rule order is important: if text does not match the first rule added to the grammar, the lexer tries to match the second rule added, and so on."

The first rule, 'myany1', is being matched, but the token style is 'default', which doesn't show anything visually distinctive. If you change the token to something like `lexer.STRING` and reset, you'll see that it works.

You can use Tools > Show Style to help investigate how a lexer is processing text.

**Reply by metaleap** [reply]
You're right! Somehow I misunderstood how `add_style` works then I guess. Let's say I want to "match-any" (got that now) and then really force `{ italics = true, bold = true, back = 0x686868 }` on all those matches, for reasons.  =)

**Reply by metaleap** [reply]
Update: got it now  =)  thx Mitchell!

```lua
local lex = lexer.new('dbgbuf')
local style = lexer.styles.default .. { italics = true, bold = true, back = 0x686868 }

lex:add_rule('myany1', token('myany1', lexer.any^1))
lex:add_style('myany1', style)
```

**Last riddle:** whole-buffer background-color (separate from other buffers with their own different lexers) isn't possible, is it?

**Comment by orbitalquark** [comment]
`lex:add_style()` only works for custom token names passed to `token()`. In your example, you are using `lexer.KEYWORD`. Try 'myany1' instead. Rule names are arbitrary and exist on their own, irrespective of tokens or styles.
